railties/lib/rails/commands/server/server_command.rb

```diff
@@ -56,10 +57,19 @@ def default_options
     end
 
     def served_url
-      "#{options[:SSLEnable] ? 'https' : 'http'}://#{options[:Host]}:#{options[:Port]}" unless use_puma?
+      server_url unless use_puma?
     end
 
     private
+      def server_url
+        "#{options[:SSLEnable] ? 'https' : 'http'}://#{options[:Host]}:#{options[:Port]}"
+      end
+
+      def write_server_url
+        path = File.join(Rails.root, "tmp", "server_url.txt")
```

Comment:
`Rails.root` supports `join`:
```suggestion
        path = Rails.root.join("tmp", "server_url.txt")
```